cpp/src/lance/encodings/binary.cc

```diff
@@ -41,12 +41,18 @@ Result<int64_t> VarBinaryEncoder::Write(const std::shared_ptr<::arrow::Array> da
   ARROW_ASSIGN_OR_RAISE(auto offsets_position, out_->Tell());
   offsets_builder_.Reset();
   assert(arr->length() > 0);
+  /// Reset the slice's first offset to zero.
+  auto first_offset = arr->value_offset(0);
   for (int64_t i = 0; i <= arr->length(); ++i) {
-    ARROW_RETURN_NOT_OK(offsets_builder_.Append(start_offset + arr->value_offset(i)));
+    ARROW_RETURN_NOT_OK(
+        offsets_builder_.Append(start_offset + arr->value_offset(i) - first_offset));
```

Comment:
What's start offset vs first offset?

Reply:
start_offset is the offset in the output file

first_offset is the "offset" of the value in the "in-memory array".

I can change these terms if confusing